frontend/src/lib/components/graph/FlowGraph.svelte

```diff
@@ -0,0 +1,226 @@
+<script lang="ts">
+	import Svelvet, { type Edge } from "svelvet";
+	import { sugiyama, dagStratify, decrossOpt } from 'd3-dag'
+	import type { FlowModule, RawScript } from "../../gen"
+	import {
+		NODE,
+		createIdGenerator,
+		isNode,
+		isLoop,
+		isBranch,
+		type GraphItem,
+		type Node,
+		type Loop,
+		type Branch,
+		type NestedNodes,
+		type ModuleHost
+	} from "."
+
+	export let modules: FlowModule[] | undefined = []
+	const idGenerator = createIdGenerator()
+	let nestedNodes: NestedNodes
+	let nodes: Node[] = []
+	let edges: Edge[] = []
+	let width: number, height: number
+	
+	$: if(modules?.length) {
+		createGraph(modules)
+	} else {
+		nodes = edges = []
+	}
+
+	function createGraph(modules: FlowModule[]) {
+		nestedNodes = nodes = []
+
+		nestedNodes.push(createVirtualNode(getParentIds(), 'Flow start'))
+		modules.forEach(m => {
+			const item = getConvertedFlowModule(m)
+			item && nestedNodes.push(item)
+		})
+		nestedNodes.push(createVirtualNode(getParentIds(), 'Flow end'))
+
+		const flatNodes = flattenNestedNodes(nestedNodes)
+		nodes = layoutNodes(flatNodes)
+		edges = createEdges(nodes)
+	}
+
+	function getConvertedFlowModule(module: FlowModule, parent: NestedNodes | undefined = undefined): GraphItem | undefined {
+		const type = module.value.type
+		const parentIds = getParentIds(parent)
+
+		if(type === 'rawscript') {
+			const lang = module.value.language
+			return flowModuleToNode(parentIds, module.id, module.summary || 'Inline script', 'inline', lang)
+		} else if(type === 'script') {
+			return flowModuleToNode(parentIds, module.id, module.summary || module.value.path || 'Hub script', 'hub')
+		} else if(type === 'forloopflow') {
+			const expr = module.value.iterator['expr']
+			return flowModuleToLoop(module.value.modules, expr ? `Expression: ${expr}` : 'For loop', parent)
+		} else if(type === 'branchone') {
+			const branches = [module.value.default, ...module.value.branches.map(b => b.modules)]
+			return flowModuleToBranch(branches, parent)
+		} else if(type === 'branchall') {
+			const branches = module.value.branches.map(b => b.modules)
+			return flowModuleToBranch(branches, parent)
+		}
+		return flowModuleToNode(parentIds, module.id, module.summary || 'Identity script', 'inline')
+	}
+
+	function getParentIds(items: NestedNodes | undefined = undefined): string[] {
+		const item = items?.at(-1) || nestedNodes.at(-1)
+		if(!item) return []
+
+		if(isNode(item)) {
+			return ['' + item.id]
```

Comment:
Are id either string of number ? I see you are also doing some: `+numberAsString` to convert things back and forth

Reply:
Svelvet library uses numbers as IDs but the layout library accepts an array of strings as the parentIds parameter so the conversion is needed to make the two packages work together